allow overriding the service worker scope to values other than rootURL

meirish opened the pull request:
Hello! Thanks for this project. I'm opening this to add the ability to override the service worker scope to something other than `rootURL`. This is useful in two situations: you want to scope the service worker more tightly, or if you want to control network requests higher up than the rootURL. 

My current use case is that I have an ember app at `/ui/`, but my api is at `/v1/`, so to interact wit h those requests, I need to have a `rooteURL` of `/ui/` and scope of `/` or `/v1/`, and then set the HTTP header `Allow-Service-Woker`  to the same value.

To accomplish this, in the PR I added a new config option - `serviceWorkerScope` - that will be used instead of `rootURL` if it's defined.

Ravenstine commented:
@meirish I like this idea and is a feature I needed a while ago.  Any chance you can add info on this to the documentation and a test?

meirish commented:
@Ravenstine 👋 - I added a blurb in the registration docs about the new setting. I tried to add a test, but noticed there's not any tests that are currently testing the rollup replacement that's happening already. It also seems that the rollup replacement isn't happening in the tests. Do you know what I need to do to get that to work (assuming we want this test)?

meirish commented:
@Ravenstine ok this is ready to go - it looks like travis isn't working, but all of the tests pass locally for me. wanted to thank @johncowen for pointing me in the right direction for getting rollup to work (it needed to be in a `ember-service-worker` folder). I felt a bit like we're just testing rollup at that point, but it did surface a bug where if you didn't provide `serviceWorkerScope`, then registration wouldn't happen properly so I'm fine leaving this in just to show the expected output if nothing else.

eshtadc commented:
Travis is resolved and the dependencies updated. Try pulling it in to see if it resolves the failed build here.

meirish commented:
@eshtadc thanks! I just rebased so hopefully tests will be 🙆‍♂

eshtadc commented on docs/_documentation/configuration.md:
````diff
@@ -77,6 +77,30 @@ In addition to `inline` strategy there is also the `async` option. This will add
+};
+```
+
+You can read more about ServiceWorker scope on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers#Enter_service_workers).
````
Can we also add a caution regarding that the scope can only _narrow_ the scope of the service worker?
For example:

> Since a service worker can't have a scope broader than its own location, only use the scope option when you need a scope that is narrower than the default.

from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/register)

meirish replied:
You did this in https://github.com/meirish/ember-service-worker/pull/1 🙌

eshtadc commented on service-worker-registration/index.js:
```diff
@@ -2,9 +2,14 @@ export const PROJECT_REVISION = '{{PROJECT_REVISION}}';
 
 let SUCCESS_HANDLERS = [];
 let ERROR_HANDLERS = [];
+const SCOPE = '{{SERVICE_WORKER_SCOPE}}';
```
Instead of doing all the work here in the registration file itself, what if we just passed in the adjusted scope? Would make testing easier as well I think.
For example, in `_rollupTree` from the `ServiceWorkerBuilder`
```
SERVICE_WORKER_SCOPE: this.options.serviceWorkerScope || this.options.rootURL
```

Then your test file only needs to verify that the scope is passed in scope or the default scope.

eshtadc commented on node-tests/fixtures/builder-test/rollup/scope-override/ember-service-worker/service-worker-registration/index.js:
```diff
@@ -0,0 +1,8 @@
+let scopePassed = SCOPE !== 'undefined';
+let scope = scopePassed ? SCOPE : ROOT_URL;
+
+if ('serviceWorker' in navigator) {
```
I'm worried that repeating the logic here in the fixture means that if the code changes (and breaks) the test will still pass. If we make the change of moving the logic into the service worker builder, then this fixture only has to test the scope passed in.... something like:
```
if ('serviceWorker' in navigator) {
  navigator.serviceWorker.register('{{ROOT_URL}}{{SERVICE_WORKER_FILENAME}}', { scope: '{{SERVICE_WORKER_SCOPE}}' });
};
```

eshtadc commented on node-tests/service-worker-builder-test.js:
```diff
@@ -127,6 +127,29 @@ describe('Service Worker Builder', () => {
         assert.equal(files[serviceWorkerFilename], expected);
       });
   });
+  it('overides root url when using serviceWorkerScope', () => {
```
This is great - can we also test that the root url is used as the scope when the scope is not explicitly passed?

meirish replied:
Just added tests for this for a custom rootURL and the default `/`.

NetForce1 commented:
I'm having this exact same issue, and I'd like to see the PR merged. What is the current status? Is anything withholding it from being merged? If so, is that anything I would be able to help with?

NoahFisher commented:
Bump on ☝️ - I have a project where this would also be useful. Is there any progress to be made on this? Happy to contribute if it's blocked by anything concrete.

meirish commented:
OK! I think I've addressed all of the feedback, thanks (and sorry) for hanging in there folks. Maybe we can squeak this in before the ... checks calendar ... two year mark?! Really, again, sorry! 😬 

Let me know if there's other things to fix up.

Thanks!

meirish commented:
@Ravenstine @eshtadc bump on this as I think it's ready to go! thanks!

hashishaw commented:
Hi there! 👋 Friendly bump, any updates on this?

Ravenstine commented:
Hi @hashishaw!  Unfortunately, neither I nor Lisa are working at DockYard anymore, so we don't have the capability of merging this.

I recommend that someone maintain a separate fork of this add-on or any other add-ons from DockYard that are still useful.  DockYard isn't focusing much attention on Ember projects anymore and I highly doubt their add-ons will see any support in the foreseeable future.
